authentication-via-social-networks/README.md

```diff
@@ -0,0 +1,726 @@
+    user.skip_confirmation!
+    user.save!
+    user
+  end
+
+  private
+
+  def user_params
+    password = Devise.friendly_token.first(8)
+    {
+      email: auth.info.email,
+      full_name: auth.info.name,
+      password: password,
+      password_confirmation: password
+    }
+  end
+end
+```
+
+```ruby
+# app/interactors/fetch_oauth_user.rb
+class FetchOauthUser
+  attr_reader :auth
+  private :auth
+
+  def initialize(auth)
+    @auth = auth
+  end
+
+  def call
+    user_found_by_uid || user_found_by_email || new_user
+  end
+
+  private
+
+  def user_found_by_uid
+    Identity.from_omniauth(auth)&.user
+  end
+
+  def user_found_by_email
+    FindUserByEmail.new(auth).call
+  end
+
+  def new_user
+    CreateUserFromAuth.new(auth).call
+  end
+end
+```
+
+```ruby
+# app/interactors/find_user_by_email.rb
+class FindUserByEmail
+  attr_reader :auth
+  private :auth
+
+  def initialize(auth)
+    @auth = auth
+  end
+
+  def call
+    return unless user
+
+    create_identity
+    user.confirm unless user.confirmed?
+    user
+  end
+
+  private
+
+  def user
+    @user ||= User.find_by(email: auth.info.email)
+  end
+
+  def create_identity
+    user.identities.where(provider: auth.provider, uid: auth.uid).first_or_create!
+  end
+end
+```
+
+```ruby
+# app/models/identity.rb
+class Identity < ActiveRecord::Base
+  PROVIDERS = OmniAuth.strategies.map { |s| s.to_s.demodulize.underscore }.drop(1)
+
+  belongs_to :user
+
+  validates :user, :provider, :uid, presence: true
+  validates :uid, uniqueness: { scope: :provider }
+
+  def self.from_omniauth(auth)
+    find_by(provider: auth.provider, uid: auth.uid)
+  end
+end
+```
+
+```ruby
+# app/models/user.rb
+class User < ActiveRecord::Base
+  devise ..., :omniauthable, omniauth_providers: Identity::PROVIDERS
+
+  has_many :identities, dependent: :destroy
+  ...
+end
+```
+
+```ruby
+# app/policies/auth_verification_policy.rb
+class AuthVerificationPolicy
+  class OauthError < StandardError
+  end
+
+  attr_reader :auth
+  private :auth
+
+  def initialize(auth)
+    @auth = auth
+  end
+
+  def verified?
+    send(auth.provider)
+  rescue NoMethodError
+    fail OauthError, I18n.t("omniauth.verification.not_implemented", kind: auth.provider)
+  end
+
+  private
+
+  def facebook
+    auth.info.verified? || auth.extra.raw_info.verified?
+  end
+
+  def google_oauth2
+    auth.extra.raw_info.email_verified?
+  end
+end
+```
+
+```ruby
+# app/views/identities/_list.html.slim
+- if identities.any?
+  b Successfully authorized via:
+  ul.js-identities
+    - identities.each do |identity|
+      li = render "identities/provider_link", identity: identity
+b Add service to sign in with:
+ul
+  - Identity::PROVIDERS.each do |provider|
+    li = link_to provider_name(provider), user_omniauth_authorize_path(provider)
+```
+
+```ruby
+# app/views/identities/_provider_link.html.slim
+= link_to "#{provider_name(identity.provider)} (#{identity.uid.truncate(9)}). Unauthorize?",
```

Comment:
Just wondering if it's a good idea - to attach multiple identities from single kind of provider, to one account?

It feels to be an excessive feature, actually :)

Also, bare text in views should be preferably replaced with I18n translations